group14's assignment1 completed

ricardodong opened the pull request:
All features has been implemented and there is .jar file for execution.

ricardodong committed:
initialization & draw detection function in borad.java added.

ricardodong committed:
added draw detection in each playmove in board.java
added the tied situation in playgame in tictactoegame.java.
so the feature1 is completed.

ricardodong committed:
Merge pull request #1 from ricardodong/feature_1_detect_tied

Feature 1 detect tied and is completed

alsobaie committed:
Added ScoreBoard Class

alsobaie committed:
ScoreBoard - added printScore function

alsobaie committed:
Scoreboard - Changed score printing style

alsobaie committed:
Merge pull request #5 from ricardodong/feature_2_scoreboard

Scoreboard - Changed score printing style [Issue feature_2 closed]

ricardodong committed:
Changed the tictactoegame to be able to run multiple times but it doesn't display score after end the game.

I left a blank function called printScore in board.java that need to be implemented.

ricardodong committed:
Merge pull request #6 from ricardodong/feature_3_play_multi_times_withour_printScore

Issue3 partially done, displaying the score remains unfinished

alsobaie committed:
Fix Issue Where ScoreBoard resets After Each Game

alsobaie committed:
Merge pull request #8 from ricardodong/Fix_ScoreBoard_Reset

Fix: ScoreBoard Resets (Completed)

ricardodong committed:
change the pom.xml

ricardodong committed:
add the assignment1.jar file that can be executed.

ricardodong committed:
change the readme to indicate the group id

ricardodong committed:
Merge pull request #9 from ricardodong/readme&pom_modification_&jar_added

Readme&pom modification &jar added

mtfische reviewed (changes requested):
### There is a game breaking bug in the user input:
   When the player enters an incorrect input in the play again feature
   the player is locked in an infinate loop.

### problem code:

   When the player is asked to enter their response again it updates the
   string variable (userInputForNextGame) but, it will never update the 
   char variable (inputchoice). https://git.io/fhR57

### steps to recreate:

1. pull latest PR code
2. mvn clean package assembly:single
3. navigate to targets/
4. java -jar TicTacToe-1.0-SNAPSHOT-jar-with-dependencies.jar
5. Complete a game
6. Enter an incorrect input 
7. This will place the input check into an infinite loop

### Severity:
   Necessary change before merge.

mtfische commented:
### A few other notable issues:
- You are not supposed to include the jar file in the repo
- The Pull Request description doesn't contain details of what the pull request has done.
- There is an inconsistent use of tabs in TicTacToeGame.java. https://git.io/fhRdg
- There is an inconsistent use of tabs in Board.java. https://git.io/fhRFg (Line 25 as well)
- ScoreBoard.java uses tabs and not spaces.
- **array** is not a descriptive variable name. https://git.io/fhRFf
- **idx** is not a descriptive parameter name. https://git.io/fhRFY

ricardodong committed:
There is a bug in tictactoegame.java that when asked to input  to play next game or not, if the input is invalid, there will be an infinite loop because the char inputchoice that is used for decision remains unchanged after new input.

It is fixed by change the inputchoice inside the loop.

ricardodong committed:
Merge pull request #10 from ricardodong/bug_fix_for_infinite_loop

infinite loop bug fixed

ricardodong committed:
Delete assignment1.jar

.jar file is not necessary

ricardodong committed:
Merge pull request #11 from ricardodong/ricardodong-delate-jarfile

Delete assignment1.jar

ricardodong commented:
> ### There is a game breaking bug in the user input:
> When the player enters an incorrect input in the play again feature
> the player is locked in an infinate loop.
> 
> ### problem code:
> When the player is asked to enter their response again it updates the
> string variable (userInputForNextGame) but, it will never update the
> char variable (inputchoice). https://git.io/fhR57
> 
> ### steps to recreate:
> 1. pull latest PR code
> 2. mvn clean package assembly:single
> 3. navigate to targets/
> 4. java -jar TicTacToe-1.0-SNAPSHOT-jar-with-dependencies.jar
> 5. Complete a game
> 6. Enter an incorrect input
> 7. This will place the input check into an infinite loop

This bug is fixed and for other code style problems, I need to talk with my group members. 
Thanks for your quick review!

Hk-tang commented on src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java:
```diff
@@ -46,11 +47,54 @@ public void playGame(){
+        if (board.getDraw()){
+        		this.board.incDraw(0);
+        		this.board.incDraw(1);
+            System.out.println("draw");
```
Should avoid debug messages in finalized code

mtfische replied:
This is not a debug message. It declares to the user that the game has ended in a draw.

Hk-tang reviewed (commented):
There are some un-intuitive variable names such as incDraw and getDraw, isDraw would be a bit more intuitive for example. There also seems to be very high coupling between Game, Board, and Scoreboard. Getters and Setters would be better than direct access.

forgeno commented on src/main/java/cs/ualberta/cmput402/tictactoe/board/ScoreBoard.java:
```diff
@@ -0,0 +1,37 @@
+		}
+	}
+	
+	public void incWin(int idx) {
```
Implementation of score is un-necessarily complicated. Please make functions more human-legible or add clear comments how the flow of the function works.

mtfische replied:
This doesn't provide enough information on what the problem is.
@forgeno try to provide: 
- Why this is unnecessarily complicated
- Input on how this implementation could be improved
- how it could be more legible/why it isn't currently

forgeno replied:
The function is unclear, please add comments or refactor code:
1. What does incWin mean? Is it an abbreviation for something?
2. paramter "idx" is not a meaningful variable name, please rename and explain what the uses are.
3. Please explain for the other 2 similar functions "incWin" and "incDraw"

mtfische replied:
This has been addressed.

forgeno commented on src/main/java/cs/ualberta/cmput402/tictactoe/board/ScoreBoard.java:
```diff
@@ -0,0 +1,37 @@
+		this.array[idx][0] += 1;
+	}
+	
+	public void incLose(int idx) {
```
Same comment as above

mtfische replied:
this doesn't expand on the issue

forgeno commented on src/main/java/cs/ualberta/cmput402/tictactoe/board/ScoreBoard.java:
```diff
@@ -0,0 +1,37 @@
+		this.array[idx][1] += 1;
+	}
+	
+	public void incDraw(int idx) {
```
Same comment as above

mtfische replied:
this doesn't expand on the issue.

mtfische reviewed (changes requested):
### Issues to address:
   Please complete or respond to the corrections outlined in Issue #3 
   The issues are also outlined in this comment: 
   https://github.com/mtfische/tictactoe/pull/2#issuecomment-455649399

### Severity:
   These are moderately sever and should be addressed before merging.
   A fix is not necessarily required.

mtfische commented:
> There are some un-intuitive variable names such as incDraw and getDraw, isDraw would be a bit more intuitive for example. There also seems to be very high coupling between Game, Board, and Scoreboard. Getters and Setters would be better than direct access.

@Hk-tang could you clarify this comment? Is isDraw a suggested solution because those are very different. Also perhaps a better solution would be to remove scoreboard from the board class and place it in TicTacToeGame instead. Let me know what you think.

Hk-tang commented:
What I meant by it is I didn't really understand what the variables were meant for and I made that suggestion upon reading the code and trying to reason what it was really doing and suggesting a better name for it. Alternatively writing a comment on variable declaration to explain its purpose would solve the issue as well. Scoreboard can remain its own class but be instantiated in TicTacToeGame if that is what you mean, which would be better if Board does not need Scoreboard.

Hk-tang reviewed (changes requested):
**Severity:**
Nice to have

Implementation should be modified slightly to instantiating the Scoreboard class in TicTacToeGame rather than Board. That way board does not need to implement methods which are just wrappers over Scoreboard's methods.
- 1) Remove Scoreboard entirely from Board 

- 2) Instantiate Scoreboard in TicTacToeGame
  - Call Scoreboard methods here directly

ricardodong committed:
I fixed the inconsistency of the use of tab and space in tictactoegame.java. I changed all the tabs to spaces.

Commit:
Fixed the inconsistency of the use of tab and space in board.java. Changed all the tabs to spaces. Also changed the name of getDraw to isDraw

Commit:
changed the use of getDraw() to isDraw(), which is more intuitive.

Commit:
Changed the output message when a game end in a draw, from "draw" to "The game end in a draw.".

ricardodong committed:
Merge pull request #12 from ricardodong/code_style_fix

Code style fix

ricardodong commented:
* @Hk-tang, @mtfische  I think that isDraw is a better name, because this function means to get the draw situation, so isDraw can be more intuitive, I would change it.
* @mtfische  I have fixed the inconsistency of tab and space use in board.java and tictactoegame.java. I changed all the tabs to spaces.  
* @mtfische  The jar file has already been delete in the previous PR.

alsobaie committed:
Refactored ScoreBoard

alsobaie commented:
Refactored _ScoreBoard_:

- _ScoreBoard_ is now an _int_ array included as an attribute in _Board_

- Names of methods _incWins_/_incLoses_/_incDraws_ for _ScoreBoard_ was 
changed to _incrementNumOfWins_/_incrementNumOfLoses_/_incrementNumOfDraws_.

- Parameter _int idx_ in the above methods was changed to _Player player_
and the logic of the methods above was changed to accommodate the new
parameter type

forgeno reviewed (approved):
Changes look good to me.
